### ISSUE TITLE
Subtypes as function parameters satisfy interface declaration

### ISSUE THREAD

**Issue by m-b-davis** [Microsoft/TypeScript#16921]
I believe this is unexpected behavior. I'm able to define an interface IConfig, with a function f accepting a parameter IBase, and then supply an IConfig with a function with a parameter which extends IBase. 

This function can now access members of IChild, even though it's being passed an IBase.

Is this behaviour expected? If so it makes my life easier, but I'm not sure why it should be valid.

**TypeScript Version:**  2.4.0 

**Code**

```ts
interface IBase { a: string; }
interface IChild extends IBase { b: number; }
interface IConfig { f: (base: IBase) => number; }

const base1 = { a: 'hi' };

const config: IConfig = {
    f: (child: IChild) => child.b
};

// base1.b is undefined
console.log(config.f(base1));
```

A simpler variation with a type instead of an interface for IConfig:

```ts
interface IBase { a: string; }
const base1 = { a: 'hi' };

interface IChild extends IBase { b: number; }
type BaseFunc = (base: IBase) => number;

// base1.b is undefined
console.log(config.f(base1));
```

**Expected behavior:**
An error compiling as f: (child: IChild) => number should not be compatible with (base:IBase) => number

**Actual behavior:**
Code complies and errors on accessing base.b

**Comment by yortus** [comment]
Function parameters are bivariant in TypeScript. That's why `tsc` does not issue an error when you pass an `IBase` to the function expecting an `IChild`, even though as you say, it fails at runtime.

This comes up fairly often, so it has its own [FAQ entry here](https://github.com/Microsoft/TypeScript/wiki/FAQ#why-are-function-parameters-bivariant).
